### PR TITLE
DEV-2408: Stripe webhooks fail for subscriptions

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This fixes a bug discovered in UAT for DEV-2183. It should be merged into DEV-2183-root-branch, and that branch will eventually be merged into main.

#### What's this PR do?

This PR fixes a bug that was causing Stripe webhooks to fail for recurring contributions ("Subscriptions" in Stripe lingo).

The root cause of the bug was that we were not saving the Stripe Customer ID back on to the NRE contribution after creating a Stripe PaymentIntent or Subscription. In our webhooks code, for subscriptions, we were attempting to retrieve by customer ID instead of subscription ID.

To solve for this, this PR changes the contribution retrieval behavior of our webhook so that it searches contributions by Stripe subscription ID for subscriptions. It also updates the payment intent and subscription creation methods in apps.contributions.models.Contribution so that the customer id gets written back to the contribution (even though the customer id is no longer relevant for the webhook in question).

#### Why are we doing this? How does it help us?

Fixes a bug that stops us from shipping DEV-2183.

#### How should this be manually tested? Please include detailed step-by-step instructions.

Create a recurring contribution ([here](https://billypenn-dev-2408.revengine-review.org/donate)), then check in NRH admin ([here](https://dev-2408.revengine-review.org/nrhadmin/contributions/)) and its status should be "paid".

In addition to fixing bug, changes a .info to a .exception so we get sentry errors when webhooks don't match with contributions.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Tests have been updated

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-2408 | Webhooks are broken for Subscriptions in DEV-2183](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-2408)
- [DEV-2183](https://news-revenue-hub.atlassian.net/browse/DEV-2183)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No

### CODE REVIEW

File: apps/contributions/models.py [hunk 1]
```diff
@@ -320,6 +321,7 @@ def create_stripe_subscription(self, stripe_customer_id, metadata):
         )
         self.payment_provider_data = subscription
         self.provider_subscription_id = subscription["id"]
+        self.provider_customer_id = subscription["customer"]
```
Comment (njharman) [inline apps/contributions/models.py:324]: It doesn't seem that test_stripe_subscription checks that this value is set.

Comment (njharman) [inline apps/contributions/models.py:324]: It doesn't seem that test_stripe_subscription checks that this value is set.

File: apps/contributions/models.py [hunk 2]
```diff
@@ -289,6 +289,7 @@ def create_stripe_one_time_payment_intent(self, stripe_customer_id, metadata):
         )
         self.provider_payment_id = intent["id"]
         self.provider_client_secret_id = intent["client_secret"]
+        self.provider_customer_id = intent["customer"]
```
Comment (njharman) [inline apps/contributions/models.py:292]: It doesn't seem that test_create_stripe_one_time_payment_intent checks that this value is set. In fact deleting this line causes no failure in test_models.py (where I'd expect failure) and no failure in any test.

File: apps/contributions/webhooks.py [hunk 3]
```diff
@@ -11,17 +11,22 @@
...
+            return Contribution.objects.get(provider_subscription_id=self.obj_data["id"])
+        elif event_type == "payment_intent":
+            try:
+                return Contribution.objects.get(provider_payment_id=self.obj_data["id"])
+            except Contribution.DoesNotExist as e:
+                if customer_id := self.obj_data.get("customer"):
+                    # This is fine as long as we continue to generate a unique customer per charge.
+                    return Contribution.objects.get(provider_customer_id=customer_id)
+                else:
+                    raise e
```
Comment (njharman) [inline apps/contributions/webhooks.py:29]: Pretty sure the "e" is superfluous and thus confusing (implying that code is modifying or raising some other exception than one being handled) https://docs.python.org/3/reference/simple_stmts.html#raise

**Review by njharman — CHANGES_REQUESTED** [review]
no time logged
no pre-review check list

File: apps/contributions/webhooks.py (hunk 3 above)
Comment (benjaminEwhite) [reply apps/contributions/webhooks.py]: b7405441

File: apps/contributions/models.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:292]: b7405441

File: apps/contributions/models.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:324]: b7405441

**Review by x110dc — APPROVED** [review]

**Pull request merged**